Topic4/src/tests/ShoppingCartTest.java

```diff
@@ -0,0 +1,201 @@
+		Item item2 = new Product("Home Theatre Sony Muteki 5.1", "HT200", 600.00);
+		Offer offer1 = new Offer("Cinema in your living", "OF001", 1150.00);
+		offer1.addItem(item1);
+		offer1.addItem(item2);
+		shopcart.addItem(offer1);
+		assertEquals(shopcart.getItems().size(), 1);
+		List<Item> items = shopcart.getItems();
+		assertEquals(offer1, items.get(0));
+
+		// product1
+		Item item3 = new Product("Table TV O", "TT001", 100.00);
+		shopcart.addItem(item3);
+		assertEquals(shopcart.getItems().size(), 2);
+
+		items = shopcart.getItems();
+		assertEquals(offer1, items.get(0));
+		assertEquals(item3, items.get(1));
+		double total = 0;
+		for (Item item : items) {
+			total += item.getValue();
+		}
+		assertTrue(total == 1250.00);
+	}
+
+	@Test
+	public void removingItems() {
+
+		ShoppingCart shopcart = new FunctionalShoppingCart();
+
+		// offer1
+		Item item1 = new Product("Smart TV LG 40 inches", "TV100", 700.00);
+		Item item2 = new Product("Home Theatre Sony Muteki 5.1", "HT200", 600.00);
+		Offer offer1 = new Offer("Cinema in your living", "OF001", 1150.00);
+		offer1.addItem(item1);
+		offer1.addItem(item2);
+		shopcart.addItem(offer1);
+
+		// product1
+		Item item3 = new Product("Table TV O", "TT001", 100.00);
+		shopcart.addItem(item3);
+
+		assertEquals(shopcart.getItems().size(), 2);
+		List<Item> items = shopcart.getItems();
+		assertEquals(offer1, items.get(0));
+		assertEquals(item3, items.get(1));
+
+		double total = 0;
+		for (Item item : items) {
+			total += item.getValue();
+		}
+		assertTrue(total == 1250.00);
+
+		shopcart.removeItem(offer1);
+		assertEquals(shopcart.getItems().size(), 1);
+		items = shopcart.getItems();
+		assertEquals(item3, items.get(0));
+
+		total = 0;
+		for (Item item : items) {
+			total += item.getValue();
+		}
+		assertTrue(total == 100.00);
+
+		shopcart.removeItem(item3);
+		assertEquals(shopcart.getItems().size(), 0);
+
+		total = 0;
+		for (Item item : items) {
+			total += item.getValue();
+		}
+		assertTrue(total == 0.00);
+	}
+
+	@Test
+	public void addingPayment() {
+		PaymentStrategy payment1 = new CashStrategy();
+		ShoppingCart shopcart = new FunctionalShoppingCart();
+		shopcart.addPayment(payment1);
+		assertEquals(shopcart.getPayment(), payment1);
+
+		PaymentStrategy payment2 = new PaypalStrategy("luciano@gmail.com", "bootcamp");
+		shopcart.addPayment(payment2);
+		assertEquals(shopcart.getPayment(), payment2);
+	}
+
+	@Test
+	public void removingPayment() {
+		PaymentStrategy payment1 = new CashStrategy();
+		ShoppingCart shopcart = new FunctionalShoppingCart();
+		shopcart.addPayment(payment1);
+		assertEquals(shopcart.getPayment(), payment1);
+		shopcart.removePayment();
+		assertEquals(shopcart.getPayment(), null);
+
+		PaymentStrategy payment2 = new CreditCardStrategy("Luciano", 11501);
+		shopcart.addPayment(payment2);
+		assertEquals(shopcart.getPayment(), payment2);
+		shopcart.removePayment();
+		assertEquals(shopcart.getPayment(), null);
+	}
+}

```

Comment:
You need to add more tests to validate the following features:
1- Create a ShoppingCart where a user can buy different items and then decide how to pay them:
- by Credit Card: where a Name and a Credit Number is required.
- by Paypal: where an email and a password is needed.
- by Cash: no information is required.

2- Add the following discount depending on the payment type
- 10% by Credit Card.
- The cheapest item is for free when the user pay by Paypal.
- 90% of the most expensive item is free if the user pays by Cash.

3- Create a counter to provide unique sequential numbers and use it to set an identification to each payment transaction.

and as you added the notification (to do this you need to implement the notification using some email service and doing the tests with some Mock Mail )

4- Add a mailist List where the market manager is notified when:
- A new item/offer is added
- A price is changed
- A new transaction was made